{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting flask==1.0.2\n",
      "  Using cached Flask-1.0.2-py2.py3-none-any.whl (91 kB)\n",
      "Collecting Werkzeug>=0.14\n",
      "  Using cached Werkzeug-2.0.2-py3-none-any.whl (288 kB)\n",
      "Collecting itsdangerous>=0.24\n",
      "  Using cached itsdangerous-2.0.1-py3-none-any.whl (18 kB)\n",
      "Requirement already satisfied: Jinja2>=2.10 in /opt/conda/lib/python3.7/site-packages (from flask==1.0.2) (2.11.1)\n",
      "Requirement already satisfied: click>=5.1 in /opt/conda/lib/python3.7/site-packages (from flask==1.0.2) (7.0)\n",
      "Requirement already satisfied: MarkupSafe>=0.23 in /opt/conda/lib/python3.7/site-packages (from Jinja2>=2.10->flask==1.0.2) (1.1.1)\n",
      "Installing collected packages: Werkzeug, itsdangerous, flask\n",
      "Successfully installed Werkzeug-2.0.2 flask-1.0.2 itsdangerous-2.0.1\n",
      "Collecting google_api_core==1.26.0\n",
      "  Using cached google_api_core-1.26.0-py2.py3-none-any.whl (92 kB)\n",
      "Requirement already satisfied: pytz in /opt/conda/lib/python3.7/site-packages (from google_api_core==1.26.0) (2019.3)\n",
      "Requirement already satisfied: setuptools>=40.3.0 in /opt/conda/lib/python3.7/site-packages (from google_api_core==1.26.0) (45.2.0.post20200209)\n",
      "Collecting google-auth<2.0dev,>=1.21.1\n",
      "  Using cached google_auth-1.35.0-py2.py3-none-any.whl (152 kB)\n",
      "Requirement already satisfied: packaging>=14.3 in /opt/conda/lib/python3.7/site-packages (from google_api_core==1.26.0) (20.1)\n",
      "Requirement already satisfied: googleapis-common-protos<2.0dev,>=1.6.0 in /opt/conda/lib/python3.7/site-packages (from google_api_core==1.26.0) (1.51.0)\n",
      "Requirement already satisfied: six>=1.13.0 in /opt/conda/lib/python3.7/site-packages (from google_api_core==1.26.0) (1.14.0)\n",
      "Collecting protobuf>=3.12.0\n",
      "  Using cached protobuf-3.19.4-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.1 MB)\n",
      "Requirement already satisfied: requests<3.0.0dev,>=2.18.0 in /opt/conda/lib/python3.7/site-packages (from google_api_core==1.26.0) (2.23.0)\n",
      "Requirement already satisfied: cachetools<5.0,>=2.0.0 in /opt/conda/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.21.1->google_api_core==1.26.0) (4.0.0)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4; python_version >= \"3.6\" in /opt/conda/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.21.1->google_api_core==1.26.0) (4.0)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /opt/conda/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.21.1->google_api_core==1.26.0) (0.2.8)\n",
      "Requirement already satisfied: pyparsing>=2.0.2 in /opt/conda/lib/python3.7/site-packages (from packaging>=14.3->google_api_core==1.26.0) (2.4.6)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/conda/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google_api_core==1.26.0) (2019.11.28)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /opt/conda/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google_api_core==1.26.0) (1.25.7)\n",
      "Requirement already satisfied: idna<3,>=2.5 in /opt/conda/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google_api_core==1.26.0) (2.9)\n",
      "Requirement already satisfied: chardet<4,>=3.0.2 in /opt/conda/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google_api_core==1.26.0) (3.0.4)\n",
      "Requirement already satisfied: pyasn1>=0.1.3 in /opt/conda/lib/python3.7/site-packages (from rsa<5,>=3.1.4; python_version >= \"3.6\"->google-auth<2.0dev,>=1.21.1->google_api_core==1.26.0) (0.4.8)\n",
      "Installing collected packages: google-auth, protobuf, google-api-core\n",
      "  Attempting uninstall: google-auth\n",
      "    Found existing installation: google-auth 1.11.2\n",
      "    Uninstalling google-auth-1.11.2:\n",
      "      Successfully uninstalled google-auth-1.11.2\n",
      "  Attempting uninstall: protobuf\n",
      "    Found existing installation: protobuf 3.11.4\n",
      "    Uninstalling protobuf-3.11.4:\n",
      "      Successfully uninstalled protobuf-3.11.4\n",
      "  Attempting uninstall: google-api-core\n",
      "    Found existing installation: google-api-core 1.16.0\n",
      "    Uninstalling google-api-core-1.16.0:\n",
      "      Successfully uninstalled google-api-core-1.16.0\n",
      "Successfully installed google-api-core-1.26.0 google-auth-1.35.0 protobuf-3.19.4\n",
      "Collecting grpcio\n",
      "  Using cached grpcio-1.43.0-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.1 MB)\n",
      "Requirement already satisfied: six>=1.5.2 in /opt/conda/lib/python3.7/site-packages (from grpcio) (1.14.0)\n",
      "Installing collected packages: grpcio\n",
      "Successfully installed grpcio-1.43.0\n",
      "Requirement already satisfied: google-cloud-core in /opt/conda/lib/python3.7/site-packages (1.3.0)\n",
      "Requirement already satisfied: google-api-core<2.0.0dev,>=1.16.0 in /opt/conda/lib/python3.7/site-packages (from google-cloud-core) (1.26.0)\n",
      "Requirement already satisfied: google-auth<2.0dev,>=1.21.1 in /opt/conda/lib/python3.7/site-packages (from google-api-core<2.0.0dev,>=1.16.0->google-cloud-core) (1.35.0)\n",
      "Requirement already satisfied: setuptools>=40.3.0 in /opt/conda/lib/python3.7/site-packages (from google-api-core<2.0.0dev,>=1.16.0->google-cloud-core) (45.2.0.post20200209)\n",
      "Requirement already satisfied: googleapis-common-protos<2.0dev,>=1.6.0 in /opt/conda/lib/python3.7/site-packages (from google-api-core<2.0.0dev,>=1.16.0->google-cloud-core) (1.51.0)\n",
      "Requirement already satisfied: requests<3.0.0dev,>=2.18.0 in /opt/conda/lib/python3.7/site-packages (from google-api-core<2.0.0dev,>=1.16.0->google-cloud-core) (2.23.0)\n",
      "Requirement already satisfied: six>=1.13.0 in /opt/conda/lib/python3.7/site-packages (from google-api-core<2.0.0dev,>=1.16.0->google-cloud-core) (1.14.0)\n",
      "Requirement already satisfied: packaging>=14.3 in /opt/conda/lib/python3.7/site-packages (from google-api-core<2.0.0dev,>=1.16.0->google-cloud-core) (20.1)\n",
      "Requirement already satisfied: pytz in /opt/conda/lib/python3.7/site-packages (from google-api-core<2.0.0dev,>=1.16.0->google-cloud-core) (2019.3)\n",
      "Requirement already satisfied: protobuf>=3.12.0 in /opt/conda/lib/python3.7/site-packages (from google-api-core<2.0.0dev,>=1.16.0->google-cloud-core) (3.19.4)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4; python_version >= \"3.6\" in /opt/conda/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.21.1->google-api-core<2.0.0dev,>=1.16.0->google-cloud-core) (4.0)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /opt/conda/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.21.1->google-api-core<2.0.0dev,>=1.16.0->google-cloud-core) (0.2.8)\n",
      "Requirement already satisfied: cachetools<5.0,>=2.0.0 in /opt/conda/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.21.1->google-api-core<2.0.0dev,>=1.16.0->google-cloud-core) (4.0.0)\n",
      "Requirement already satisfied: idna<3,>=2.5 in /opt/conda/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-api-core<2.0.0dev,>=1.16.0->google-cloud-core) (2.9)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/conda/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-api-core<2.0.0dev,>=1.16.0->google-cloud-core) (2019.11.28)\n",
      "Requirement already satisfied: chardet<4,>=3.0.2 in /opt/conda/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-api-core<2.0.0dev,>=1.16.0->google-cloud-core) (3.0.4)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /opt/conda/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-api-core<2.0.0dev,>=1.16.0->google-cloud-core) (1.25.7)\n",
      "Requirement already satisfied: pyparsing>=2.0.2 in /opt/conda/lib/python3.7/site-packages (from packaging>=14.3->google-api-core<2.0.0dev,>=1.16.0->google-cloud-core) (2.4.6)\n",
      "Requirement already satisfied: pyasn1>=0.1.3 in /opt/conda/lib/python3.7/site-packages (from rsa<5,>=3.1.4; python_version >= \"3.6\"->google-auth<2.0dev,>=1.21.1->google-api-core<2.0.0dev,>=1.16.0->google-cloud-core) (0.4.8)\n",
      "Processing /home/jovyan/.cache/pip/wheels/50/9d/19/dc1a639905431252dbf8a61388a83459c5cfef7add445bebe4/gcloud-0.18.3-py3-none-any.whl\n",
      "Requirement already satisfied: httplib2>=0.9.1 in /opt/conda/lib/python3.7/site-packages (from gcloud) (0.17.0)\n",
      "Requirement already satisfied: six in /opt/conda/lib/python3.7/site-packages (from gcloud) (1.14.0)\n",
      "Requirement already satisfied: protobuf!=3.0.0.b2.post1,>=3.0.0b2 in /opt/conda/lib/python3.7/site-packages (from gcloud) (3.19.4)\n",
      "Requirement already satisfied: googleapis-common-protos in /opt/conda/lib/python3.7/site-packages (from gcloud) (1.51.0)\n",
      "Requirement already satisfied: oauth2client>=2.0.1 in /opt/conda/lib/python3.7/site-packages (from gcloud) (4.1.3)\n",
      "Requirement already satisfied: pyasn1>=0.1.7 in /opt/conda/lib/python3.7/site-packages (from oauth2client>=2.0.1->gcloud) (0.4.8)\n",
      "Requirement already satisfied: rsa>=3.1.4 in /opt/conda/lib/python3.7/site-packages (from oauth2client>=2.0.1->gcloud) (4.0)\n",
      "Requirement already satisfied: pyasn1-modules>=0.0.5 in /opt/conda/lib/python3.7/site-packages (from oauth2client>=2.0.1->gcloud) (0.2.8)\n",
      "Installing collected packages: gcloud\n",
      "Successfully installed gcloud-0.18.3\n",
      "Collecting google-api-python-client==1.7.8\n",
      "  Using cached google_api_python_client-1.7.8-py3-none-any.whl (56 kB)\n",
      "Requirement already satisfied: six<2dev,>=1.6.1 in /opt/conda/lib/python3.7/site-packages (from google-api-python-client==1.7.8) (1.14.0)\n",
      "Collecting google-auth-httplib2>=0.0.3\n",
      "  Using cached google_auth_httplib2-0.1.0-py2.py3-none-any.whl (9.3 kB)\n",
      "Requirement already satisfied: httplib2<1dev,>=0.9.2 in /opt/conda/lib/python3.7/site-packages (from google-api-python-client==1.7.8) (0.17.0)\n",
      "Requirement already satisfied: google-auth>=1.4.1 in /opt/conda/lib/python3.7/site-packages (from google-api-python-client==1.7.8) (1.35.0)\n",
      "Collecting uritemplate<4dev,>=3.0.0\n",
      "  Using cached uritemplate-3.0.1-py2.py3-none-any.whl (15 kB)\n",
      "Requirement already satisfied: cachetools<5.0,>=2.0.0 in /opt/conda/lib/python3.7/site-packages (from google-auth>=1.4.1->google-api-python-client==1.7.8) (4.0.0)\n",
      "Requirement already satisfied: setuptools>=40.3.0 in /opt/conda/lib/python3.7/site-packages (from google-auth>=1.4.1->google-api-python-client==1.7.8) (45.2.0.post20200209)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /opt/conda/lib/python3.7/site-packages (from google-auth>=1.4.1->google-api-python-client==1.7.8) (0.2.8)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4; python_version >= \"3.6\" in /opt/conda/lib/python3.7/site-packages (from google-auth>=1.4.1->google-api-python-client==1.7.8) (4.0)\n",
      "Requirement already satisfied: pyasn1<0.5.0,>=0.4.6 in /opt/conda/lib/python3.7/site-packages (from pyasn1-modules>=0.2.1->google-auth>=1.4.1->google-api-python-client==1.7.8) (0.4.8)\n",
      "Installing collected packages: google-auth-httplib2, uritemplate, google-api-python-client\n",
      "Successfully installed google-api-python-client-1.7.8 google-auth-httplib2-0.1.0 uritemplate-3.0.1\n",
      "Collecting google-cloud-kms==2.2.0\n",
      "  Using cached google_cloud_kms-2.2.0-py2.py3-none-any.whl (80 kB)\n",
      "Collecting proto-plus>=1.4.0\n",
      "  Using cached proto_plus-1.19.9-py3-none-any.whl (45 kB)\n",
      "Processing /home/jovyan/.cache/pip/wheels/b9/ee/67/2e444183030cb8d31ce8b34cee34a7afdbd3ba5959ea846380/grpc_google_iam_v1-0.12.3-py3-none-any.whl\n",
      "Collecting libcst>=0.2.5\n",
      "  Using cached libcst-0.4.1-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (2.7 MB)\n",
      "Requirement already satisfied: google-api-core[grpc]<2.0.0dev,>=1.22.2 in /opt/conda/lib/python3.7/site-packages (from google-cloud-kms==2.2.0) (1.26.0)\n",
      "Requirement already satisfied: protobuf>=3.19.0 in /opt/conda/lib/python3.7/site-packages (from proto-plus>=1.4.0->google-cloud-kms==2.2.0) (3.19.4)\n",
      "Requirement already satisfied: googleapis-common-protos[grpc]<2.0.0dev,>=1.5.2 in /opt/conda/lib/python3.7/site-packages (from grpc-google-iam-v1<0.13dev,>=0.12.3->google-cloud-kms==2.2.0) (1.51.0)\n",
      "Requirement already satisfied: grpcio<2.0.0dev,>=1.0.0 in /opt/conda/lib/python3.7/site-packages (from grpc-google-iam-v1<0.13dev,>=0.12.3->google-cloud-kms==2.2.0) (1.43.0)\n",
      "Collecting typing-extensions>=3.7.4.2\n",
      "  Using cached typing_extensions-4.0.1-py3-none-any.whl (22 kB)\n",
      "Requirement already satisfied: pyyaml>=5.2 in /opt/conda/lib/python3.7/site-packages (from libcst>=0.2.5->google-cloud-kms==2.2.0) (5.3)\n",
      "Collecting typing-inspect>=0.4.0\n",
      "  Using cached typing_inspect-0.7.1-py3-none-any.whl (8.4 kB)\n",
      "Requirement already satisfied: pytz in /opt/conda/lib/python3.7/site-packages (from google-api-core[grpc]<2.0.0dev,>=1.22.2->google-cloud-kms==2.2.0) (2019.3)\n",
      "Requirement already satisfied: requests<3.0.0dev,>=2.18.0 in /opt/conda/lib/python3.7/site-packages (from google-api-core[grpc]<2.0.0dev,>=1.22.2->google-cloud-kms==2.2.0) (2.23.0)\n",
      "Requirement already satisfied: six>=1.13.0 in /opt/conda/lib/python3.7/site-packages (from google-api-core[grpc]<2.0.0dev,>=1.22.2->google-cloud-kms==2.2.0) (1.14.0)\n",
      "Requirement already satisfied: google-auth<2.0dev,>=1.21.1 in /opt/conda/lib/python3.7/site-packages (from google-api-core[grpc]<2.0.0dev,>=1.22.2->google-cloud-kms==2.2.0) (1.35.0)\n",
      "Requirement already satisfied: packaging>=14.3 in /opt/conda/lib/python3.7/site-packages (from google-api-core[grpc]<2.0.0dev,>=1.22.2->google-cloud-kms==2.2.0) (20.1)\n",
      "Requirement already satisfied: setuptools>=40.3.0 in /opt/conda/lib/python3.7/site-packages (from google-api-core[grpc]<2.0.0dev,>=1.22.2->google-cloud-kms==2.2.0) (45.2.0.post20200209)\n",
      "Collecting mypy-extensions>=0.3.0\n",
      "  Using cached mypy_extensions-0.4.3-py2.py3-none-any.whl (4.5 kB)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/conda/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-api-core[grpc]<2.0.0dev,>=1.22.2->google-cloud-kms==2.2.0) (2019.11.28)\n",
      "Requirement already satisfied: chardet<4,>=3.0.2 in /opt/conda/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-api-core[grpc]<2.0.0dev,>=1.22.2->google-cloud-kms==2.2.0) (3.0.4)\n",
      "Requirement already satisfied: idna<3,>=2.5 in /opt/conda/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-api-core[grpc]<2.0.0dev,>=1.22.2->google-cloud-kms==2.2.0) (2.9)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /opt/conda/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-api-core[grpc]<2.0.0dev,>=1.22.2->google-cloud-kms==2.2.0) (1.25.7)\n",
      "Requirement already satisfied: cachetools<5.0,>=2.0.0 in /opt/conda/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.21.1->google-api-core[grpc]<2.0.0dev,>=1.22.2->google-cloud-kms==2.2.0) (4.0.0)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /opt/conda/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.21.1->google-api-core[grpc]<2.0.0dev,>=1.22.2->google-cloud-kms==2.2.0) (0.2.8)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4; python_version >= \"3.6\" in /opt/conda/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.21.1->google-api-core[grpc]<2.0.0dev,>=1.22.2->google-cloud-kms==2.2.0) (4.0)\n",
      "Requirement already satisfied: pyparsing>=2.0.2 in /opt/conda/lib/python3.7/site-packages (from packaging>=14.3->google-api-core[grpc]<2.0.0dev,>=1.22.2->google-cloud-kms==2.2.0) (2.4.6)\n",
      "Requirement already satisfied: pyasn1<0.5.0,>=0.4.6 in /opt/conda/lib/python3.7/site-packages (from pyasn1-modules>=0.2.1->google-auth<2.0dev,>=1.21.1->google-api-core[grpc]<2.0.0dev,>=1.22.2->google-cloud-kms==2.2.0) (0.4.8)\n",
      "Installing collected packages: proto-plus, grpc-google-iam-v1, typing-extensions, mypy-extensions, typing-inspect, libcst, google-cloud-kms\n",
      "Successfully installed google-cloud-kms-2.2.0 grpc-google-iam-v1-0.12.3 libcst-0.4.1 mypy-extensions-0.4.3 proto-plus-1.19.9 typing-extensions-4.0.1 typing-inspect-0.7.1\n",
      "Requirement already satisfied: pyodbc in /opt/conda/lib/python3.7/site-packages (4.0.30)\n",
      "Collecting firebase_admin\n",
      "  Using cached firebase_admin-5.2.0-py3-none-any.whl (115 kB)\n",
      "Collecting google-cloud-storage>=1.37.1\n",
      "  Using cached google_cloud_storage-2.1.0-py2.py3-none-any.whl (106 kB)\n",
      "Requirement already satisfied: google-api-core[grpc]<3.0.0dev,>=1.22.1; platform_python_implementation != \"PyPy\" in /opt/conda/lib/python3.7/site-packages (from firebase_admin) (1.26.0)\n",
      "Requirement already satisfied: google-api-python-client>=1.7.8 in /opt/conda/lib/python3.7/site-packages (from firebase_admin) (1.7.8)\n",
      "Collecting google-cloud-firestore>=2.1.0; platform_python_implementation != \"PyPy\"\n",
      "  Using cached google_cloud_firestore-2.3.4-py2.py3-none-any.whl (232 kB)\n",
      "Collecting cachecontrol>=0.12.6\n",
      "  Using cached CacheControl-0.12.10-py2.py3-none-any.whl (20 kB)\n",
      "Collecting google-resumable-media>=1.3.0\n",
      "  Using cached google_resumable_media-2.1.0-py2.py3-none-any.whl (75 kB)\n",
      "Requirement already satisfied: requests<3.0.0dev,>=2.18.0 in /opt/conda/lib/python3.7/site-packages (from google-cloud-storage>=1.37.1->firebase_admin) (2.23.0)\n",
      "Requirement already satisfied: protobuf in /opt/conda/lib/python3.7/site-packages (from google-cloud-storage>=1.37.1->firebase_admin) (3.19.4)\n",
      "Collecting google-cloud-core<3.0dev,>=1.6.0\n",
      "  Using cached google_cloud_core-2.2.2-py2.py3-none-any.whl (29 kB)\n",
      "Requirement already satisfied: google-auth<3.0dev,>=1.25.0 in /opt/conda/lib/python3.7/site-packages (from google-cloud-storage>=1.37.1->firebase_admin) (1.35.0)\n",
      "Requirement already satisfied: setuptools>=40.3.0 in /opt/conda/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.22.1; platform_python_implementation != \"PyPy\"->firebase_admin) (45.2.0.post20200209)\n",
      "Requirement already satisfied: six>=1.13.0 in /opt/conda/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.22.1; platform_python_implementation != \"PyPy\"->firebase_admin) (1.14.0)\n",
      "Requirement already satisfied: pytz in /opt/conda/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.22.1; platform_python_implementation != \"PyPy\"->firebase_admin) (2019.3)\n",
      "Requirement already satisfied: googleapis-common-protos<2.0dev,>=1.6.0 in /opt/conda/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.22.1; platform_python_implementation != \"PyPy\"->firebase_admin) (1.51.0)\n",
      "Requirement already satisfied: packaging>=14.3 in /opt/conda/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.22.1; platform_python_implementation != \"PyPy\"->firebase_admin) (20.1)\n",
      "Requirement already satisfied: grpcio<2.0dev,>=1.29.0; extra == \"grpc\" in /opt/conda/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.22.1; platform_python_implementation != \"PyPy\"->firebase_admin) (1.43.0)\n",
      "Requirement already satisfied: uritemplate<4dev,>=3.0.0 in /opt/conda/lib/python3.7/site-packages (from google-api-python-client>=1.7.8->firebase_admin) (3.0.1)\n",
      "Requirement already satisfied: httplib2<1dev,>=0.9.2 in /opt/conda/lib/python3.7/site-packages (from google-api-python-client>=1.7.8->firebase_admin) (0.17.0)\n",
      "Requirement already satisfied: google-auth-httplib2>=0.0.3 in /opt/conda/lib/python3.7/site-packages (from google-api-python-client>=1.7.8->firebase_admin) (0.1.0)\n",
      "Requirement already satisfied: proto-plus>=1.10.0 in /opt/conda/lib/python3.7/site-packages (from google-cloud-firestore>=2.1.0; platform_python_implementation != \"PyPy\"->firebase_admin) (1.19.9)\n",
      "Requirement already satisfied: msgpack>=0.5.2 in /opt/conda/lib/python3.7/site-packages (from cachecontrol>=0.12.6->firebase_admin) (1.0.0)\n",
      "Collecting google-crc32c<2.0dev,>=1.0\n",
      "  Using cached google_crc32c-1.3.0-cp37-cp37m-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (38 kB)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/conda/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-cloud-storage>=1.37.1->firebase_admin) (2019.11.28)\n",
      "Requirement already satisfied: idna<3,>=2.5 in /opt/conda/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-cloud-storage>=1.37.1->firebase_admin) (2.9)\n",
      "Requirement already satisfied: chardet<4,>=3.0.2 in /opt/conda/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-cloud-storage>=1.37.1->firebase_admin) (3.0.4)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /opt/conda/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-cloud-storage>=1.37.1->firebase_admin) (1.25.7)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4; python_version >= \"3.6\" in /opt/conda/lib/python3.7/site-packages (from google-auth<3.0dev,>=1.25.0->google-cloud-storage>=1.37.1->firebase_admin) (4.0)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /opt/conda/lib/python3.7/site-packages (from google-auth<3.0dev,>=1.25.0->google-cloud-storage>=1.37.1->firebase_admin) (0.2.8)\n",
      "Requirement already satisfied: cachetools<5.0,>=2.0.0 in /opt/conda/lib/python3.7/site-packages (from google-auth<3.0dev,>=1.25.0->google-cloud-storage>=1.37.1->firebase_admin) (4.0.0)\n",
      "Requirement already satisfied: pyparsing>=2.0.2 in /opt/conda/lib/python3.7/site-packages (from packaging>=14.3->google-api-core[grpc]<3.0.0dev,>=1.22.1; platform_python_implementation != \"PyPy\"->firebase_admin) (2.4.6)\n",
      "Requirement already satisfied: pyasn1>=0.1.3 in /opt/conda/lib/python3.7/site-packages (from rsa<5,>=3.1.4; python_version >= \"3.6\"->google-auth<3.0dev,>=1.25.0->google-cloud-storage>=1.37.1->firebase_admin) (0.4.8)\n",
      "\u001b[31mERROR: google-cloud-storage 2.1.0 has requirement google-api-core<3.0dev,>=1.29.0, but you'll have google-api-core 1.26.0 which is incompatible.\u001b[0m\n",
      "\u001b[31mERROR: google-cloud-bigquery 1.24.0 has requirement google-cloud-core<2.0dev,>=1.1.0, but you'll have google-cloud-core 2.2.2 which is incompatible.\u001b[0m\n",
      "\u001b[31mERROR: google-cloud-bigquery 1.24.0 has requirement google-resumable-media<0.6dev,>=0.5.0, but you'll have google-resumable-media 2.1.0 which is incompatible.\u001b[0m\n",
      "Installing collected packages: google-crc32c, google-resumable-media, google-cloud-core, google-cloud-storage, google-cloud-firestore, cachecontrol, firebase-admin\n",
      "  Attempting uninstall: google-resumable-media\n",
      "    Found existing installation: google-resumable-media 0.5.0\n",
      "    Uninstalling google-resumable-media-0.5.0:\n",
      "      Successfully uninstalled google-resumable-media-0.5.0\n",
      "  Attempting uninstall: google-cloud-core\n",
      "    Found existing installation: google-cloud-core 1.3.0\n",
      "    Uninstalling google-cloud-core-1.3.0:\n",
      "      Successfully uninstalled google-cloud-core-1.3.0\n",
      "  Attempting uninstall: google-cloud-storage\n",
      "    Found existing installation: google-cloud-storage 1.26.0\n",
      "    Uninstalling google-cloud-storage-1.26.0:\n",
      "      Successfully uninstalled google-cloud-storage-1.26.0\n",
      "Successfully installed cachecontrol-0.12.10 firebase-admin-5.2.0 google-cloud-core-2.2.2 google-cloud-firestore-2.3.4 google-cloud-storage-2.1.0 google-crc32c-1.3.0 google-resumable-media-2.1.0\n",
      "Requirement already satisfied: google-cloud-firestore in /opt/conda/lib/python3.7/site-packages (2.3.4)\n",
      "Requirement already satisfied: google-cloud-core<3.0.0dev,>=1.4.1 in /opt/conda/lib/python3.7/site-packages (from google-cloud-firestore) (2.2.2)\n",
      "Requirement already satisfied: google-api-core[grpc]<3.0.0dev,>=1.26.0 in /opt/conda/lib/python3.7/site-packages (from google-cloud-firestore) (1.26.0)\n",
      "Requirement already satisfied: proto-plus>=1.10.0 in /opt/conda/lib/python3.7/site-packages (from google-cloud-firestore) (1.19.9)\n",
      "Requirement already satisfied: packaging>=14.3 in /opt/conda/lib/python3.7/site-packages (from google-cloud-firestore) (20.1)\n",
      "Requirement already satisfied: google-auth<3.0dev,>=1.24.0 in /opt/conda/lib/python3.7/site-packages (from google-cloud-core<3.0.0dev,>=1.4.1->google-cloud-firestore) (1.35.0)\n",
      "Requirement already satisfied: requests<3.0.0dev,>=2.18.0 in /opt/conda/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.26.0->google-cloud-firestore) (2.23.0)\n",
      "Requirement already satisfied: setuptools>=40.3.0 in /opt/conda/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.26.0->google-cloud-firestore) (45.2.0.post20200209)\n",
      "Requirement already satisfied: six>=1.13.0 in /opt/conda/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.26.0->google-cloud-firestore) (1.14.0)\n",
      "Requirement already satisfied: protobuf>=3.12.0 in /opt/conda/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.26.0->google-cloud-firestore) (3.19.4)\n",
      "Requirement already satisfied: googleapis-common-protos<2.0dev,>=1.6.0 in /opt/conda/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.26.0->google-cloud-firestore) (1.51.0)\n",
      "Requirement already satisfied: pytz in /opt/conda/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.26.0->google-cloud-firestore) (2019.3)\n",
      "Requirement already satisfied: grpcio<2.0dev,>=1.29.0; extra == \"grpc\" in /opt/conda/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.26.0->google-cloud-firestore) (1.43.0)\n",
      "Requirement already satisfied: pyparsing>=2.0.2 in /opt/conda/lib/python3.7/site-packages (from packaging>=14.3->google-cloud-firestore) (2.4.6)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4; python_version >= \"3.6\" in /opt/conda/lib/python3.7/site-packages (from google-auth<3.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google-cloud-firestore) (4.0)\n",
      "Requirement already satisfied: cachetools<5.0,>=2.0.0 in /opt/conda/lib/python3.7/site-packages (from google-auth<3.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google-cloud-firestore) (4.0.0)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /opt/conda/lib/python3.7/site-packages (from google-auth<3.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google-cloud-firestore) (0.2.8)\n",
      "Requirement already satisfied: chardet<4,>=3.0.2 in /opt/conda/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-api-core[grpc]<3.0.0dev,>=1.26.0->google-cloud-firestore) (3.0.4)\n",
      "Requirement already satisfied: idna<3,>=2.5 in /opt/conda/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-api-core[grpc]<3.0.0dev,>=1.26.0->google-cloud-firestore) (2.9)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/conda/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-api-core[grpc]<3.0.0dev,>=1.26.0->google-cloud-firestore) (2019.11.28)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /opt/conda/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-api-core[grpc]<3.0.0dev,>=1.26.0->google-cloud-firestore) (1.25.7)\n",
      "Requirement already satisfied: pyasn1>=0.1.3 in /opt/conda/lib/python3.7/site-packages (from rsa<5,>=3.1.4; python_version >= \"3.6\"->google-auth<3.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google-cloud-firestore) (0.4.8)\n",
      "Processing /home/jovyan/.cache/pip/wheels/dc/9a/e9/49e627353476cec8484343c4ab656f1e0d783ee77b9dde2d1f/crcmod-1.7-cp37-cp37m-linux_x86_64.whl\n",
      "Installing collected packages: crcmod\n",
      "Successfully installed crcmod-1.7\n",
      "Collecting pandas==1.3.5\n",
      "  Using cached pandas-1.3.5-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (11.3 MB)\n",
      "Requirement already satisfied: numpy>=1.17.3; platform_machine != \"aarch64\" and platform_machine != \"arm64\" and python_version < \"3.10\" in /opt/conda/lib/python3.7/site-packages (from pandas==1.3.5) (1.18.1)\n",
      "Requirement already satisfied: python-dateutil>=2.7.3 in /opt/conda/lib/python3.7/site-packages (from pandas==1.3.5) (2.8.1)\n",
      "Requirement already satisfied: pytz>=2017.3 in /opt/conda/lib/python3.7/site-packages (from pandas==1.3.5) (2019.3)\n",
      "Requirement already satisfied: six>=1.5 in /opt/conda/lib/python3.7/site-packages (from python-dateutil>=2.7.3->pandas==1.3.5) (1.14.0)\n",
      "Installing collected packages: pandas\n",
      "  Attempting uninstall: pandas\n",
      "    Found existing installation: pandas 1.0.1\n",
      "    Uninstalling pandas-1.0.1:\n",
      "      Successfully uninstalled pandas-1.0.1\n",
      "Successfully installed pandas-1.3.5\n",
      "Collecting virtualenv\n",
      "  Using cached virtualenv-20.13.0-py2.py3-none-any.whl (6.5 MB)\n",
      "Requirement already satisfied: importlib-metadata>=0.12; python_version < \"3.8\" in /opt/conda/lib/python3.7/site-packages (from virtualenv) (1.5.0)\n",
      "Collecting filelock<4,>=3.2\n",
      "  Using cached filelock-3.4.2-py3-none-any.whl (9.9 kB)\n",
      "Requirement already satisfied: six<2,>=1.9.0 in /opt/conda/lib/python3.7/site-packages (from virtualenv) (1.14.0)\n",
      "Collecting platformdirs<3,>=2\n",
      "  Using cached platformdirs-2.4.1-py3-none-any.whl (14 kB)\n",
      "Collecting distlib<1,>=0.3.1\n",
      "  Using cached distlib-0.3.4-py2.py3-none-any.whl (461 kB)\n",
      "Requirement already satisfied: zipp>=0.5 in /opt/conda/lib/python3.7/site-packages (from importlib-metadata>=0.12; python_version < \"3.8\"->virtualenv) (3.0.0)\n",
      "Installing collected packages: filelock, platformdirs, distlib, virtualenv\n",
      "Successfully installed distlib-0.3.4 filelock-3.4.2 platformdirs-2.4.1 virtualenv-20.13.0\n",
      "Requirement already satisfied: XlsxWriter in /opt/conda/lib/python3.7/site-packages (1.2.8)\n"
     ]
    }
   ],
   "source": [
    "!pip install flask==1.0.2\n",
    "!pip install google_api_core==1.26.0\n",
    "!pip install grpcio\n",
    "!pip install google-cloud-core\n",
    "\n",
    "!pip install gcloud\n",
    "!pip install google-api-python-client==1.7.8\n",
    "!pip install google-cloud-kms==2.2.0\n",
    "\n",
    "!pip install pyodbc\n",
    "!pip install firebase_admin\n",
    "!pip install google-cloud-firestore\n",
    "!pip install crcmod==1.7\n",
    "\n",
    "!pip install pandas==1.3.5\n",
    "!pip install virtualenv\n",
    "\n",
    "!pip install XlsxWriter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import firebase_admin\n",
    "from firebase_admin import credentials\n",
    "from firebase_admin import firestore\n",
    "\n",
    "from datetime import timedelta, time\n",
    "import pickle\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import json\n",
    "import base64\n",
    "import math\n",
    "\n",
    "import xlsxwriter\n",
    "import smtplib\n",
    "import gspread\n",
    "import re\n",
    "import io\n",
    "import tempfile\n",
    "import pyodbc\n",
    "import time\n",
    "\n",
    "import tempfile\n",
    "\n",
    "from google.cloud import bigquery as bq\n",
    "from google.oauth2 import service_account\n",
    "from google.cloud.exceptions import NotFound\n",
    "from google.cloud import storage\n",
    "\n",
    "from email.mime.multipart import MIMEMultipart\n",
    "from email.mime.text import MIMEText\n",
    "from email.mime.base import MIMEBase\n",
    "from email import encoders\n",
    "\n",
    "from genericFunctions2 import SQLconnection\n",
    "from functionEncription2 import decryptCatalogs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# conecction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('credenciales.json','r') as f:\n",
    "    mailSettings = json.load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# enviar informacion por correo automatico"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "email_1@gmail.com\n",
      "email_2@gmail.com\n",
      "Nombre del correo\n"
     ]
    }
   ],
   "source": [
    "# pipeline_parameters = json.loads(parameters)\n",
    "hiddenMail = \"email_3@gmail.com\"\n",
    "\n",
    "group = 'grupo'\n",
    "emailBody = \"Nombre del correo\"\n",
    "toMail = (',').join([\"email_1@gmail.com\"])\n",
    "ccMail = (',').join([\"email_2@gmail.com\"])\n",
    "\n",
    "print(toMail)\n",
    "print(ccMail)\n",
    "print(emailBody)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_1=pd.DataFrame()\n",
    "df_2=pd.DataFrame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(221, b'Bye')"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_a = df_1.copy()\n",
    "df_b = df_2.copy()\n",
    "\n",
    "output = io.BytesIO()\n",
    "\n",
    "# Use the BytesIO object as the filehandle.\n",
    "writer = pd.ExcelWriter(output, engine='xlsxwriter')\n",
    "\n",
    "# Write the data frame to the BytesIO object.\n",
    "df_a.to_excel(writer, sheet_name='nombre hoja 1', startrow=2, header=False, index=False)\n",
    "df_b.to_excel(writer, sheet_name='nombre hoja 2', startrow=2, header=False, index=False)\n",
    "workbook  = writer.book\n",
    "\n",
    "#---------------------------------------------\n",
    "worksheet = writer.sheets['nombre hoja 1']\n",
    "\n",
    "# Add formats.\n",
    "header_format = workbook.add_format({'bold': True, 'text_wrap': True, 'valign': 'top', 'align': 'center', 'fg_color': '#4e0000' , 'color': '#ffffff', 'border': 1})\n",
    "title_format = workbook.add_format({'bold': True, 'text_wrap': True, 'valign': 'top', 'align': 'left', 'font_size': 16, 'fg_color': '#ffffff' , 'color': '#4e0000'})\n",
    "cell_format = workbook.add_format({'align': 'center', 'border': 1, 'border_color': '#000000'}) \n",
    "\n",
    "# Write the column headers with the defined format.\n",
    "for col_num , value in enumerate(dfTrips.columns.values):\n",
    "    worksheet.write(1, col_num , value, header_format)\n",
    "\n",
    "title = 'TITULO HOJA 1'\n",
    "worksheet.merge_range('A1:Q1', title, title_format)\n",
    "worksheet.set_column(0, 16, 25, cell_format)\n",
    "\n",
    "# writer.save()\n",
    "#-------------------------------------------\n",
    "worksheet = writer.sheets['nombre hoja 2']\n",
    "\n",
    "# Add formats.\n",
    "header_format = workbook.add_format({'bold': True, 'text_wrap': True, 'valign': 'top', 'align': 'center', 'fg_color': '#4e0000' , 'color': '#ffffff', 'border': 1})\n",
    "title_format = workbook.add_format({'bold': True, 'text_wrap': True, 'valign': 'top', 'align': 'left', 'font_size': 16, 'fg_color': '#ffffff' , 'color': '#4e0000'})\n",
    "cell_format = workbook.add_format({'align': 'center', 'border': 1, 'border_color': '#000000'}) \n",
    "\n",
    "# Write the column headers with the defined format.\n",
    "for col_num , value in enumerate(df.columns.values):\n",
    "    worksheet.write(2, col_num , value, header_format)\n",
    "\n",
    "title = 'TITULO HOJA 2'\n",
    "worksheet.merge_range('A1:Q1', title, title_format)\n",
    "worksheet.set_column(0, 16, 25, cell_format)\n",
    "\n",
    "writer.save()\n",
    "#-------------------------------------------\n",
    "\n",
    "xlsx_data = output.getvalue()\n",
    "\n",
    "# Send file\n",
    "smtpHostName = mailSettings['a']['HostName']\n",
    "smtpPort =     mailSettings['a']['Port']\n",
    "smtpUserName = mailSettings['a']['UserName']\n",
    "smtpPassword = mailSettings['a']['Password']\n",
    "fromMail = \"sender@gmail.com\"\n",
    "\n",
    "msg = MIMEMultipart('alternative')\n",
    "msg['Subject'] = 'nombre_archivo.xlsx'\n",
    "msg['From'] = fromMail\n",
    "msg['To'] = toMail\n",
    "msg['Cc'] = ccMail\n",
    "\n",
    "text = 'nombre_archivo.xlsx'\n",
    "fileName = 'nombre_archivo.xlsx'\n",
    "fileName = fileName.replace(\" \", \"_\")\n",
    "part1 = MIMEText(text, 'plain')\n",
    "msg.attach(part1)\n",
    "\n",
    "p = MIMEBase('application', 'octet-stream') \n",
    "p.set_payload(xlsx_data) \n",
    "encoders.encode_base64(p) \n",
    "p.add_header('Content-Disposition', \"attachment; filename= %s\" % fileName) \n",
    "msg.attach(p) \n",
    "\n",
    "# Send via local SMTP server.\n",
    "s = smtplib.SMTP(smtpHostName,smtpPort)\n",
    "s.starttls()\n",
    "s.ehlo()\n",
    "s.login(smtpUserName,smtpPassword)\n",
    "s.sendmail(fromMail, toMail.split(',') + ccMail.split(',') + hiddenMail.split(','), msg.as_string())\n",
    "s.quit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
